pkg/hub/kubectl_ui.go

```diff
@@ -0,0 +1,194 @@
+	flags.StringVar(&k.serverKey, "server-key", k.serverKey, "Secure communication with this key")
+	flags.IntVar(&k.serverPort, "server-port", k.serverPort, "handle user request using this port")
+}
+
+func (k *KubectlUI) Validate() error {
+	if k.serverCert == "" {
+		return fmt.Errorf("The server-cert is required")
+	}
+
+	if k.serverKey == "" {
+		return fmt.Errorf("The server-key is required")
+	}
+
+	if k.serverPort == 0 {
+		return fmt.Errorf("The server-port is required")
+	}
+
+	return nil
+}
+
+func NewKubectlUI() *KubectlUI {
+	return &KubectlUI{}
+}
+
+func (k *KubectlUI) init(ctx context.Context) error {
+	proxyTLSCfg, err := util.GetClientTLSConfig(k.proxyCACertPath, k.proxyCertPath, k.proxyKeyPath, k.proxyServerHost, nil)
+	if err != nil {
+		return err
+	}
+	k.getTunnel = func() (konnectivity.Tunnel, error) {
+		// instantiate a gprc proxy dialer
+		tunnel, err := konnectivity.CreateSingleUseGrpcTunnel(
+			ctx,
+			net.JoinHostPort(k.proxyServerHost, strconv.Itoa(k.proxyServerPort)),
+			grpc.WithTransportCredentials(grpccredentials.NewTLS(proxyTLSCfg)),
+		)
+		if err != nil {
+			return nil, err
+		}
+		return tunnel, nil
+	}
+	return nil
+}
+
+func (k *KubectlUI) handler(wr http.ResponseWriter, req *http.Request) {
+	if klog.V(4).Enabled() {
+		dump, err := httputil.DumpRequest(req, true)
+		if err != nil {
+			http.Error(wr, err.Error(), http.StatusBadRequest)
+			return
+		}
+		klog.V(4).Infof("request:\n%s", string(dump))
+	}
+
+	// parse clusterID from current requestURL
+	clusterID, kubeAPIPath, err := parseRequestURL(req.RequestURI)
+	if err != nil {
+		http.Error(wr, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	target := fmt.Sprintf("https://%s", clusterID)
+	apiserverURL, err := url.Parse(target)
+	if err != nil {
+		http.Error(wr, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	tunnel, err := k.getTunnel()
+	if err != nil {
+		http.Error(wr, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	var proxyConn net.Conn
+	defer func() {
+		if proxyConn != nil {
+			err = proxyConn.Close()
+			if err != nil {
+				klog.Errorf("connection closed: %v", err)
+			}
+		}
+	}()
+
+	proxy := httputil.NewSingleHostReverseProxy(apiserverURL)
+	proxy.Transport = &http.Transport{
+		Proxy:                 http.ProxyFromEnvironment,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true, // Skip server-auth for kube-apiserver
+		},
+		// golang http pkg automaticly upgrade http connection to http2 connection, but http2 can not upgrade to SPDY which used in "kubectl exec".
+		// set ForceAttemptHTTP2 = false to prevent auto http2 upgration
+		ForceAttemptHTTP2: false,
+		DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			proxyConn, err = tunnel.DialContext(ctx, network, addr)
```

Comment:
I wonder if we should maintain a long running connection here. For instance, you could have a controller that establish the connection based on the status of addon resources, and put it in the cache. Will that enhance performance?

Reply:
Yes, this is also a part of the next plan.